Reference for stage 3 arraybuffer-transfer (#28381)
mdn/content: files/en-us/web/javascript/reference/global_objects/arraybuffer/index.md

### Transferring ArrayBuffers

`ArrayBuffer` objects can be transferred between different execution contexts, like [Web Workers](/en-US/docs/Web/API/Web_Workers_API) or [Service Workers](/en-US/docs/Web/API/Service_Worker_API), using the [structured clone algorithm](/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm). This is done by passing the `ArrayBuffer` as a [transferable object](/en-US/docs/Web/API/Web_Workers_API/Transferable_objects) in a call to {{domxref("Worker.postMessage()")}} or {{domxref("ServiceWorker.postMessage()")}}. In pure JavaScript, you can also transfer the ownership of memory from one `ArrayBuffer` to another using its {{jsxref("ArrayBuffer/transfer", "transfer()")}} or {{jsxref("ArrayBuffer/transferToFixedLength", "transferToFixedLength()")}} method.

When an `ArrayBuffer` is transferred, its original copy becomes _detached_ — this means it is no longer usable. At any moment, there will only be one copy of the `ArrayBuffer` that actually has access to the underlying memory. Detached buffers have the following behaviors:

> - Its (and its typed array views') {{jsxref("ArrayBuffer/byteLength", "byteLength")}} becomes 0.
> - Its methods, such as {{jsxref("ArrayBuffer/resize", "resize()")}} and {{jsxref("ArrayBuffer/slice", "slice()")}}, throw a {{jsxref("TypeError")}} when invoked. Its typed array views' methods also throw a `TypeError`.

Reviewer: For some reason the "Its" on the bullets is annoying me:

```suggestion
- {{jsxref("ArrayBuffer/byteLength", "byteLength")}} becomes 0 (in both the array and typed array views).
- Methods, such as {{jsxref("ArrayBuffer/resize", "resize()")}} and {{jsxref("ArrayBuffer/slice", "slice()")}}, throw a {{jsxref("TypeError")}} when invoked. The associated typed array view methods also throw a `TypeError`.
```

  - : The size, in bytes, of the `ArrayBuffer`. This is established when the array is constructed and can only be changed using the {{jsxref("ArrayBuffer.prototype.resize()")}} method if the `ArrayBuffer` is resizable.
- {{jsxref("Object/constructor", "ArrayBuffer.prototype.constructor")}}
  - : The constructor function that created the instance object. For `ArrayBuffer` instances, the initial value is the {{jsxref("ArrayBuffer/ArrayBuffer", "ArrayBuffer")}} constructor.

> - {{jsxref("ArrayBuffer.prototype.detached")}}

Author: ```suggestion
- {{jsxref("ArrayBuffer.prototype.detached")}} {{Experimental_Inline}}
```

  - : Resizes the `ArrayBuffer` to the specified size, in bytes.
- {{jsxref("ArrayBuffer.prototype.slice()")}}
  - : Returns a new `ArrayBuffer` whose contents are a copy of this `ArrayBuffer`'s bytes from `begin` (inclusive) up to `end` (exclusive). If either `begin` or `end` is negative, it refers to an index from the end of the array, as opposed to from the beginning.

> - {{jsxref("ArrayBuffer.prototype.transfer()")}}

Author: ```suggestion
- {{jsxref("ArrayBuffer.prototype.transfer()")}} {{Experimental_Inline}}
```

>   - : Creates a new `ArrayBuffer` with the same byte content as this buffer, preserving its resizability, then detaches this buffer.

Reviewer: "preserving its resizability," is awkward to read. I think I would omit this here.

Reviewer: ```suggestion
  - : Creates a new `ArrayBuffer` with the same byte content as this buffer, then detaches this buffer.
```

> - {{jsxref("ArrayBuffer.prototype.transferToFixedLength()")}}

Author: ```suggestion
- {{jsxref("ArrayBuffer.prototype.transferToFixedLength()")}} {{Experimental_Inline}}
```
